syntax = "proto3";
message ProjectRequest {
  bytes requestJson = 1;
}

message Query {
  int32 id = 1;
}

message ProjectResponse {
  int32 code = 1;
  string msg = 2;
  bytes data = 3;
}

message PermissionDto {
  int32 user_id = 1;
  optional int32 user_role = 2;
  int32 project_id = 3;
}

message PermissionResponseDto {
  int32 code = 1;
  bool data = 2;
  string msg = 3;
}

message ListProjectDto {
  optional int32 page = 1;
  optional int32 size = 2;
  optional string name = 3;
}

message ProjectModel {
  int32 id = 1;
  string created_at = 2;
  string updated_at = 3;
  int32 create_user = 4;
  int32 update_user = 5;
  string name = 6;
  int32 owner = 7;
  bool private = 8;
  string description = 9;
  string dingtalk_url = 10;
  string app = 11;
}

message ListProjectResponseData {
  repeated ProjectModel data = 1;
  int32 total = 2;
}

message ListProjectResponseDto {
  int32 code = 1;
  ListProjectResponseData data = 2;
  string msg = 3;
}

message ProjectDto {
  optional int32 id = 7;
  string name = 1;
  string app = 2;
  int32 owner = 3;
  optional bool private = 4;
  string description = 5;
  string dingtalk_url = 6;
  optional string avatar = 12;
  optional string created_at = 11;
  optional string updated_at = 8;
  optional int32 create_user = 9;
  optional int32 update_user = 10;
}

message ProjectRoleDto {
  optional int32 id = 1;
  int32 user_id = 2;
  int32 project_id = 3;
  optional int32 project_role = 4;
  optional string created_at = 5;
  optional string updated_at = 6;
  optional int32 create_user = 7;
  optional int32 update_user = 8;
}

message QueryProjectData {
  ProjectDto project = 1;
  repeated ProjectRoleDto roles = 2;

}

message QueryProjectResponseDto {
  int32 code = 1;
  QueryProjectData data = 2;
  string msg = 3;
}

message QueryUserProjectResponse {
  int32 code = 1;
  string msg = 2;
  repeated int32 data = 3;
}

message ProjectAvatarDto {
  int32 project_id = 1;
  string filename = 2;
  string content = 3;
}

message ProjectAvatarResponseDto {
  int32 code = 1;
  string data = 2;
  string msg = 3;
}

message QueryUserProjectDto {
  int32 user_id = 1;
}

message QueryUserProjectAmountResponse {
  int32 code = 1;
  string msg = 2;
  int32 data = 3;
}

service project {
  // 获取项目列表
  rpc list(ListProjectDto) returns (ListProjectResponseDto) {}
  // 创建项目
  rpc insert(ProjectDto) returns (ProjectResponse) {}
  // 更新项目头像
  rpc updateAvatar(ProjectAvatarDto) returns (ProjectAvatarResponseDto) {}
  // 更新项目
  rpc update(ProjectDto) returns (ProjectResponse) {}
  // 查询项目信息
  rpc query(Query) returns (QueryProjectResponseDto) {}
  // 删除项目
  rpc delete(Query) returns (ProjectResponse) {}
  // 插入角色
  rpc insertRole(ProjectRoleDto) returns (ProjectResponse) {}
  // 编辑角色
  rpc updateRole(ProjectRoleDto) returns (ProjectResponse) {}
  // 删除角色
  rpc deleteRole(Query) returns (ProjectResponse) {}
  // 查询用户是否有权限
  rpc checkPermission(PermissionDto) returns (PermissionResponseDto) {}
  // 查询用户已有的项目数量
  rpc queryUserProjectAmount(ProjectRequest) returns (QueryUserProjectAmountResponse) {}
  // 查询用户已有的项目
  rpc queryUserProject(ProjectRequest) returns (QueryUserProjectResponse) {}
}